onnxscript/function_libs/torch_aten/ops/core.py

```diff
@@ -269,44 +269,80 @@ def aten_any(self: TensorType) -> TensorType:
 
-    # Cast input to double if dtype is specified, because the input dtype may be e.g. bool
-    # which Range does not support. The output type is ensured because the output
-    # is casted to the specified dtype.
-    if dtype != -1:
-        end = op.Cast(end, to=DOUBLE.dtype)
+    # NOTE: trace_only because both if branches need to be the same type, but we have
+    # a cast in the if branch.
 
-    result = op.Range(0, end, 1)
-    if dtype != -1:
-        result = op.Cast(result, to=dtype)
+    if _range_supported(dtype):
+        end = op.Cast(end, to=dtype)
+        zero = op.Cast(0, to=dtype)
+        one = op.Cast(1, to=dtype)
+        result = op.Range(zero, end, one)
+    else:
+        # Cast input to float if dtype is not supported by Range,
+        # because the input dtype may be e.g. bool
+        # which Range does not support. The output type is ensured because the output
+        # is casted to the specified dtype.
+        end_ = end
+        end = op.Cast(end, to=FLOAT.dtype)
+        zero = op.Constant(value_float=0.0)
+        one = op.Constant(value_float=1.0)
+        result = op.Range(zero, end, one)
+        if dtype != -1:
+            result = op.Cast(result, to=dtype)
+        else:
+            result = op.CastLike(result, end_)
 
     return result
 
 
-@torch_op("aten::arange", overload=True)
+@torch_op("aten::arange", overload=True, trace_only=True)
 def aten_arange_start(
     start: TRealUnlessFloat16OrInt8, end: TRealUnlessFloat16OrInt8, dtype: int = -1
 ) -> TensorType:
     # arange.start(Scalar start, Scalar end, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
 
-    # Cast input to double if dtype is specified, because the input dtype may be e.g. bool
-    # which Range does not support. The output type is ensured because the output
-    # is casted to the specified dtype.
-    if dtype != -1:
-        start = op.Cast(start, to=DOUBLE.dtype)
-        end = op.Cast(end, to=DOUBLE.dtype)
+    # NOTE: trace_only because both if branches need to be the same type, but we have
+    # a cast in the if branch.
 
-    result = op.Range(start, end, 1)
-    if dtype != -1:
-        result = op.Cast(result, to=dtype)
+    if _range_supported(dtype):
+        end = op.Cast(end, to=dtype)
+        start = op.Cast(start, to=dtype)
+        one = op.Cast(1, to=dtype)
+        result = op.Range(start, end, one)
+    else:
+        # Cast input to float if dtype is not supported by Range,
+        # because the input dtype may be e.g. bool
+        # which Range does not support. The output type is ensured because the output
+        # is casted to the specified dtype.
+        end_ = end
+        end = op.Cast(end, to=FLOAT.dtype)
+        start = op.Cast(start, to=FLOAT.dtype)
+        one = op.Constant(value_float=1.0)
+        result = op.Range(start, end, one)
+        if dtype != -1:
+            result = op.Cast(result, to=dtype)
+        else:
```

Comment:
In this case, it looks like all input types are supported by range. Special-casing for `dtype == -1` may be useful.

Reply:
Done